src/material/core/datetime/date-selection-model.ts

```diff
@@ -41,6 +48,34 @@ export interface DateSelectionModelChange<S> {
   source: unknown;
 }
 
+/**
+ * Function that determines how a date range is changed when the user selects a date.
+ * Has to return a new `DateRange` that may or may not include the selected date.
+ */
+export type DateRangeSelectionStrategy<D> =
+  (selectedDate: D | null, currentRange: DateRange<D>, adapter: DateAdapter<D>) => DateRange<D>;
+
+/** Token that can be used to customize the date range selection strategy. */
+export const DATE_RANGE_SELECTION_STRATEGY =
+    new InjectionToken<DateRangeSelectionStrategy<unknown>>('DATE_RANGE_SELECTION_STRATEGY', {
+      providedIn: 'root',
+      factory: () => (selectedDate: unknown | null, currentRange: DateRange<unknown>,
```

Comment:
I'm wondering if this is powerful enough, it might be good to think through some of the use cases:

**Input modes:**
- selecting with mouse (hover/click)
- selecting with touch (drag/drop)
- selecting with keyboard (arrow/enter keys)

**UI:**
- selected range
- preview (if I'm hovering/dragging/arrowing, what would happen if I clicked/dropped/entered)

I know we don't have all of these things implemented at the moment, but it would be good if our strategy could handle all of these. It seems like one crucial piece of information that's missing is "what was the event that triggered this?"

Reply:
I'm not sure what the use case would be for changing the selection behavior between mouse, touch and keyboard. The function gets invoked at the same time no matter what. Even if we did attach the source of the selection, I don't think that consumers would be able to do much with it since it only happens when the user has clicked/tapped on something.

Reply:
Hmm, in my head I think I was imagining that this function has to somehow distinguish between whether we're showing a range preview or an actual selected range. Though thinking more about it, that can probably be handled outside of this function. This function just needs to behave the same way in either case and then the consuming logic determines whether to style it as a preview or actual selected range.

The preview functionality is something I'd like to get in (as a separate PR) before rolling this out, because I have concerns about the usability of the touch and keyboard interactions without it

Reply:
By preview functionality, do you mean the UI that we should when you click on a start date and hover away? That part works when you're selecting with the keyboard as well, but it doesn't handle touch devices.

Reply:
Yeah, ok maybe just touch needs work then. I feel like touch could have special logic though which is why I was thinking about taking the event into this function. (e.g. what if I choose a range, re-open the calendar, and then drag one of the ends?)